File: ConcurrencySample/app/src/main/java/com/example/concurrency/MainActivity.java
package com.example.concurrency;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;

import java.util.concurrent.atomic.AtomicInteger;

import static java.lang.Thread.sleep;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

	private Button anrButton;
	private final AtomicInteger counter = new AtomicInteger();

	@Override
	protected void onCreate( Bundle savedInstanceState )
	{
		super.onCreate( savedInstanceState );
		setContentView( R.layout.activity_main );

		anrButton = findViewById( R.id.anrButton );
		anrButton.setOnClickListener( this );
		findViewById( R.id.kotlinButton ).setOnClickListener( this );
		findViewById( R.id.handlerAndHandlerButton ).setOnClickListener( this );
		findViewById( R.id.asyncTaskButton ).setOnClickListener( this );
		findViewById( R.id.asyncTaskAndLifecycleButton ).setOnClickListener( this );
		findViewById( R.id.loaderButton ).setOnClickListener( this );
		final Runnable updateText = new Runnable() {
			@Override
			public void run()
			{
				counter.incrementAndGet();
				anrButton.postDelayed( new Runnable() {
					@Override
					public void run()
					{
						anrButton.setText( "Changed text " + counter );
					}
				}, 1000L );
			}
		};
		try {
			sleep(5000L);
		} catch( InterruptedException e ) {
			e.printStackTrace();
		}

		Thread launcherThread = new Thread() {
			@Override
			public void run()
			{
				counter.set( 6 );
				try {
					sleep(2000);
				} catch( InterruptedException e ) {
					e.printStackTrace();
				}
				new Thread(updateText).start();
				try {
					sleep(12000);
				} catch( InterruptedException e ) {
					e.printStackTrace();
				}
				new Thread(updateText).start();
			}
		};
		launcherThread.start();
		HandlerThread handlerThread = new HandlerThread( "background" );
	}

	@Override
	public void onClick( final View v )
	{
		Intent intent;
		int id = v.getId();
		if( id == R.id.anrButton ) {
			intent = AnrActivity.getIntent( this );
		} else if( id == R.id.kotlinButton ) {
			intent = new Intent( this, KotlinActivity.class );
		} else if( id == R.id.handlerAndHandlerButton ) {
			intent = HandlerAndHandlerActivity.getIntent( this );
		} else if( id == R.id.asyncTaskButton ) {
			intent = AsyncTaskActivity.getIntent( this );
		} else {
			throw new IllegalStateException();
		}

		startActivity( intent );
	}
}

File: DataStoringSample/app/src/main/java/com/example/datastoring/MainActivity.kt
package com.example.datastoring

import android.os.AsyncTask
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.view.View
import android.widget.TextView
import com.example.datastoring.App.Companion.databaseHolder
import com.example.datastoring.db.Person
import com.example.datastoring.db.PersonRepository
import java.util.*

class MainActivity : AppCompatActivity(), View.OnClickListener {
	private lateinit var textView: TextView
	private lateinit var personRepository: PersonRepository
	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_main)
		textView = findViewById(R.id.textView)
		findViewById<View>(R.id.createButton).setOnClickListener(this)
		findViewById<View>(R.id.loadButton).setOnClickListener(this)
		personRepository = PersonRepository(databaseHolder ?: return)
	}

	override fun onClick(v: View) {
		val id = v.id
		if (id == R.id.createButton) {
			onCreateButtonClick()
		} else if (id == R.id.loadButton) {
			onLoadButtonClick()
		}
	}

	private fun onCreateButtonClick() {
		object : AsyncTask<Void?, Void?, Void?>() {
			override fun doInBackground(vararg voids: Void?): Void? {
				val person = Person()
				person.name = UUID.randomUUID().toString()
				personRepository.create(person)
				return null
			}
		}.execute()
	}

	private fun onLoadButtonClick() {
		object : AsyncTask<Void?, Void?, String>() {
			override fun doInBackground(vararg voids: Void?): String {
				return personRepository.loadAll().toString()
			}

			override fun onPostExecute(s: String) {
				super.onPostExecute(s)
				textView.text = s
			}
		}.execute()
	}
}